File: mazesolver.py
import base64
import png
import astar

robot_radius = 1
threshold = 30

def generate_obstacles(picture)
    png_reader = png.Reader(bytes=picture)
    width, height, values, info = png_reader.read_flat()
    ox, oy = [], []

    for i in range(0, width * height * 3, 3):
        if (values[i] < threshold and 
           values[i+1] < threshold and 
           values[i+2] < threshold):
            normed = i / 3
            oy.append(normed / width)
            ox.append(normed % width)

    return ox, oy

def solve_maze(picture_b64, start, goal):
    picture = base64.b64decode(picture_b64)
    ox, oy = generate_obstacles(picture)
    sx = start[0]
    sy = start[1]
    gx = goal[0]
    gy = goal[1]
    grid_size = 10

    astar_planner = a_star.AStarPlanner(ox, oy, grid_size, robot_radius)
    rx, ry = astar_planner.planning(sx, sy, gx, gy)

    return rx, ry

File: mazesolver.ts
var createPlanner = require('l1-path-finder');
var ndarray = require('ndarray');

var PNG = require('pngjs').PNG;

export function base64PictureToPicture(base64Picture: string) {
    return Buffer.from(base64Picture, 'base64');
}

export function planPath(startX: number, startY: number, goalX: number, goalY: number, maze: any) {
    if (startX < 0 || startX >= maze.shape[0] ||
        startY < 0 || startY >= maze.shape[1]) {
            throw { status: 400, message: 'start point outside of picture'}
    }
    if (goalX < 0 || goalX >= maze.shape[0] ||
        goalY < 0 || goalY >= maze.shape[1]) {
            throw { status: 400, message: 'goal point outside of picture'}
    }

    // Create path planner
    var planner = createPlanner(maze);

    // Find path
    var path = [];
    var dist = planner.search(startX, startY, goalX, goalY, path);

    if (!path.length) {
        throw { status: 400, message: 'no path found'}
    }

    var pathInfo = {'path': path, 'dist': dist};

    return pathInfo;
}

export function pngPictureToArray(pngPicture: Buffer, threshold: number) {
    try {
        var png = PNG.sync.read(pngPicture);
    } catch(err) {
        throw { status: 415, message: 'picture is not a png picture'};
    }

    var picArray = ndarray(new Array(png.width * png.height), [png.width, png.height])

    for (var y = 0; y < png.height; y++) {
        for (var x = 0; x < png.width; x++) {  
            var idx = png.width * 4 * y + x * 4;
            if (png.data[idx] < threshold &&
                png.data[idx+1] < threshold &&
                png.data[idx+2] < threshold) {
                picArray.set(x, y, 1)
            } else {
                picArray.set(x, y, 0)
            }
        }
    }

    return picArray;
}

export function solveMaze(event: any) {
    if (event.threshold === undefined) {
        event.threshold = 123;
    }
    if (event.threshold < 0 || event.threshold > 255) {
        throw { status: 400, message: 'threshold out of bounce. Must be between 0 and 255' };
    }
    var picture = base64PictureToPicture(event.data);
    if (event.datatype == 'png') {
        var picArray = pngPictureToArray(picture, event.threshold);
    } else {
        throw { status: 415, message: 'datatype not supported. Supported formats: png'}
    }
    var pathInfo = planPath(event.start.x, event.start.y, event.goal.x, event.goal.y, picArray);
    return pathInfo;
}